File: tests/oracle/fspec_possibleparam_1204.cc
/*
 * FSPEC-POSSIBLEPARAM-1204: locked Ghidra 12.0.4 oracle for the
 * FSPEC-POSSIBLEPARAM-JOIN-0006 fix (A46 residual 3: the spaceless
 * justifiedContain callers assumedExtension cc:366 and
 * ParamListStandardOut::possibleParam cc:1765).
 *
 *  - case=plain_out_entries: ParamListStandardOut::possibleParam
 *    (fspec.cc:1765-1774) iterates the RAW entry list — no resolver
 *    window, no caller-level space filter, no minSize gate — and
 *    accepts ANY non-negative justifiedContain (>= 0, not the
 *    findEntry just==0 gate), so unjustified sub-ranges are possible
 *    params. Space rejection happens only INSIDE justifiedContain:
 *    fspec.cc:269 (alignment!=0) and address.cc:133 (alignment==0).
 *  - case=join_out_entries: join entries ARE reachable: the per-piece
 *    join walk (fspec.cc:251-262) returns the containing offset
 *    (0x104/4 -> 4 >= 0 -> true), unlike findEntry(just=true) which
 *    requires 0. The old Rugra caller-level space filter made every
 *    join entry unreachable — the core divergence pinned here.
 *  - case=cross_space_join: pieces in different spaces with equal
 *    offsets; the ram query walks the foreign low piece
 *    (address.cc:133 -1, +4 skip) into the containing high piece
 *    (offset 4 >= 0) -> true.
 *  - case=assumed_extension: ParamListStandard::assumedExtension
 *    (fspec.cc:1426) -> ParamEntry::assumedExtension (fspec.cc:366):
 *    the cc:377 justifiedContain is space-aware (foreign-space query
 *    at a numerically justified offset -> CPUI_COPY), joins return
 *    CPUI_COPY at cc:376 before the containment check, the sz gates
 *    (cc:370-375), the list-level minSize skip (cc:1431), and both
 *    container shapes (whole alignment cc:383-388, whole exclusion
 *    entry cc:378-382).
 *
 * ParamEntry/ParamListStandardOut/JoinRecord construction uses the
 * same class->struct access hack as the fspec_findentry_1204 fixture;
 * every observation goes through public members only. Neither
 * possibleParam nor assumedExtension touches the resolver, so
 * populateResolver is intentionally NOT called.
 */

#include <bits/stdc++.h>

#define class struct
#define private public
#define protected public
#include "fspec.hh"
#include "opcodes.hh"
#include "space.hh"
#include "translate.hh"
#undef protected
#undef private
#undef class

using namespace ghidra;

class FixtureTranslate final : public Translate {
public:
  FixtureTranslate() {
    setBigEndian(false);
    setUniqueBase(0);
  }

  std::string tryInsertSpace(AddrSpace *spc) {
    try {
      insertSpace(spc);
      return "ok";
    }
    catch (LowlevelError &err) {
      return std::string("err ") + err.explain;
    }
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const std::string &) const override {
    static VarnodeData dummy;
    return dummy;
  }
  std::string getRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  std::string getExactRegisterName(AddrSpace *, uintb, int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData, std::string> &) const override {}
  void getUserOpNames(std::vector<std::string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &, const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &, const Address &) const override { return 0; }
};

// canonical synthetic spaces: constant=0, other=1, unique=2, ram=3,
// register=4, stack=5, join=6 (same index plan as the
// fspec_findentry_1204 fixture).
static void buildSpaces(FixtureTranslate *tr, AddrSpace *&constant,
                        AddrSpace *&unique, AddrSpace *&ram,
                        AddrSpace *&reg, AddrSpace *&stack, AddrSpace *&join) {
  tr->tryInsertSpace(new ConstantSpace(tr, tr));
  tr->tryInsertSpace(new OtherSpace(tr, tr, OtherSpace::INDEX));
  tr->tryInsertSpace(new UniqueSpace(tr, tr, 2, 0));
  ram = new AddrSpace(tr, tr, IPTR_PROCESSOR, "ram", false, 8, 1, 3,
                      AddrSpace::hasphysical, 0, 0);
  reg = new AddrSpace(tr, tr, IPTR_PROCESSOR, "register", false, 8, 1, 4,
                      AddrSpace::hasphysical, 0, 0);
  stack = new AddrSpace(tr, tr, IPTR_PROCESSOR, "stack", false, 8, 1, 5,
                        AddrSpace::hasphysical, 0, 0);
  tr->tryInsertSpace(ram);
  tr->tryInsertSpace(reg);
  tr->tryInsertSpace(stack);
  constant = tr->getConstantSpace();
  unique = tr->getUniqueSpace();
  join = new JoinSpace(tr, tr, 6);
}

// Fill ParamEntry fields directly (staged cspec loader, same pattern
// as the fspec_findentry_1204 fixture).
static ParamEntry makeEntry(int4 grp, AddrSpace *spc, uintb base, int4 size,
                            int4 minsize, int4 alignment, uint4 flags) {
  ParamEntry e(grp);
  e.flags = flags;
  e.type = TYPECLASS_GENERAL;
  e.groupSet.clear();
  e.groupSet.push_back(grp);
  e.spaceid = spc;
  e.addressbase = base;
  e.size = size;
  e.minsize = minsize;
  e.alignment = alignment;
  e.numslots = (alignment != 0) ? (size / alignment) : 1;
  e.joinrec = (JoinRecord *)0;
  return e;
}

// Stage a join ParamEntry (spaceid = join space, pieces MOST
// significant first), the state resolveJoin caches after decoding a
// join <pentry>.
static ParamEntry makeJoinEntry(int4 grp, AddrSpace *joinspc, JoinRecord *rec,
                                int4 size, int4 minsize) {
  ParamEntry e(grp);
  e.flags = 0;
  e.type = TYPECLASS_GENERAL;
  e.groupSet.clear();
  e.groupSet.push_back(grp);
  e.spaceid = joinspc;
  e.addressbase = 0;
  e.size = size;
  e.minsize = minsize;
  e.alignment = 0;
  e.numslots = 1;
  e.joinrec = rec;
  return e;
}

static void printPp(const ParamListStandardOut *out, const char *spcname,
                    AddrSpace *spc, uintb off, int4 sz) {
  bool hit = out->possibleParam(Address(spc, off), sz);
  std::cout << "  pp spc=" << spcname << " off=0x" << std::hex << off
            << " sz=" << std::dec << sz << " -> " << (hit ? "true" : "false")
            << std::endl;
}

static void printAe(const ParamListStandard *stds, const char *spcname,
                    AddrSpace *spc, uintb off, int4 sz) {
  VarnodeData res;
  res.space = (AddrSpace *)0;
  res.offset = 0;
  res.size = 0;
  OpCode ext = stds->assumedExtension(Address(spc, off), sz, res);
  std::cout << "  ae spc=" << spcname << " off=0x" << std::hex << off
            << " sz=" << std::dec << sz << " -> " << get_opname(ext);
  if (ext != CPUI_COPY)
    std::cout << " res=" << res.space->getName() << ":0x" << std::hex
              << res.offset << "/" << std::dec << res.size;
  std::cout << std::endl;
}

int main(void) {
  std::cout << "schema=1|fixture=FSPEC-POSSIBLEPARAM-1204|"
               "oracle=e40ed13014025f82488b1f8f7bca566894ac376b"
            << std::endl;

  FixtureTranslate tr;
  AddrSpace *constant, *unique, *ram, *reg, *stack, *join;
  buildSpaces(&tr, constant, unique, ram, reg, stack, join);

  // ---- case 1: plain output entries ------------------------------
  std::cout << "case=plain_out_entries" << std::endl;
  {
    ParamListStandardOut out;
    // e0: register [0x200,0x207] min 4, exclusion, force-left
    out.entry.push_back(makeEntry(0, reg, 0x200, 8, 4, 0,
                                  ParamEntry::force_left_justify));
    // e1: register [0x1000,0x100F] min 4, alignment 8 (aligned route)
    out.entry.push_back(makeEntry(1, reg, 0x1000, 16, 4, 8, 0));
    // e2: ram [0x2000,0x200F] min 2, exclusion, force-left
    out.entry.push_back(makeEntry(2, ram, 0x2000, 16, 2, 0,
                                  ParamEntry::force_left_justify));
    out.numgroup = 3;
    struct PpCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const PpCase pp[] = {
      { "reg", reg, 0x200, 4 },    // justified hit -> true
      { "reg", reg, 0x202, 2 },    // unjustified offset 2: >= 0 -> true
      { "reg", reg, 0x208, 4 },    // above e0, below e1 -> false
      { "reg", reg, 0x1000, 8 },   // aligned justified hit -> true
      { "reg", reg, 0x1002, 4 },   // aligned unjustified (2) -> true
      { "stack", stack, 0x1000, 8 }, // cc:269 foreign-space guard -> false
      { "stack", stack, 0x200, 4 },  // address.cc:133 route -> false
      { "ram", ram, 0x2000, 1 },   // NO minSize gate (min 2 > 1) -> true
      { "reg", reg, 0x200, 1 },    // NO minSize gate (min 4 > 1) -> true
      { "unique", unique, 0x200, 4 }, // foreign space -> false
    };
    for (const PpCase &c : pp)
      printPp(&out, c.spcname, c.spc, c.off, c.sz);
  }

  // ---- case 2: join entries ARE reachable --------------------------
  std::cout << "case=join_out_entries" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = reg; high.offset = 0x104; high.size = 4;
    low.space = reg; low.offset = 0x100; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandardOut out;
    // e0: join reg:0x104 (high) + reg:0x100 (low)
    out.entry.push_back(makeJoinEntry(0, join, &rec, 8, 4));
    // e1: plain register [0x200,0x207] min 4
    out.entry.push_back(makeEntry(1, reg, 0x200, 8, 4, 0,
                                  ParamEntry::force_left_justify));
    out.numgroup = 2;
    struct PpCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const PpCase pp[] = {
      { "reg", reg, 0x100, 4 },    // low piece offset 0 -> true (join!)
      { "reg", reg, 0x104, 4 },    // walk offset 4 >= 0 -> true (join!)
      { "reg", reg, 0x100, 8 },    // join walk -1, e1 no -> false
      { "reg", reg, 0x102, 4 },    // pokes out of both pieces -> false
      { "reg", reg, 0x200, 4 },    // plain entry hit -> true
      { "stack", stack, 0x100, 4 }, // per-piece address.cc:133 -> false
    };
    for (const PpCase &c : pp)
      printPp(&out, c.spcname, c.spc, c.off, c.sz);
  }

  // ---- case 3: cross-space join pieces -----------------------------
  std::cout << "case=cross_space_join" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = ram; high.offset = 0x200; high.size = 4;
    low.space = reg; low.offset = 0x200; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandardOut out;
    out.entry.push_back(makeJoinEntry(0, join, &rec, 8, 4));
    out.numgroup = 1;
    struct PpCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const PpCase pp[] = {
      { "reg", reg, 0x200, 4 },    // low piece offset 0 -> true
      { "ram", ram, 0x200, 4 },    // foreign low -1 (+4), high cur=0 -> 4 >= 0 -> true
      { "reg", reg, 0x204, 4 },    // outside both pieces -> false
      { "reg", reg, 0x200, 8 },    // low pokes out, high foreign -> -1 -> false
      { "stack", stack, 0x200, 4 }, // both pieces foreign -> false
    };
    for (const PpCase &c : pp)
      printPp(&out, c.spcname, c.spc, c.off, c.sz);
  }

  // ---- case 4: assumedExtension gates and containers ----------------
  std::cout << "case=assumed_extension" << std::endl;
  {
    JoinRecord rec;
    VarnodeData high, low;
    high.space = reg; high.offset = 0x204; high.size = 4;
    low.space = reg; low.offset = 0x200; low.size = 4;
    rec.pieces.push_back(high);
    rec.pieces.push_back(low);
    ParamListStandard stds;
    // e0: register [0x100,0x11F] min 2 alignment 8, smallsize zext
    stds.entry.push_back(makeEntry(0, reg, 0x100, 32, 2, 8,
                                   ParamEntry::smallsize_zext));
    // e1: join reg:0x204 + reg:0x200 min 2, smallsize sext (never
    // extends: cc:376 join guard)
    stds.entry.push_back(makeJoinEntry(1, join, &rec, 8, 2));
    stds.entry.back().flags = ParamEntry::smallsize_sext;
    // e2: ram [0x2000,0x200F] min 2, exclusion, smallsize inttype
    stds.entry.push_back(makeEntry(2, ram, 0x2000, 16, 2, 0,
                                   ParamEntry::smallsize_inttype));
    // e3: register [0x3000,0x3007] min 2, exclusion, smallsize sext
    stds.entry.push_back(makeEntry(3, reg, 0x3000, 8, 2, 0,
                                   ParamEntry::smallsize_sext));
    // e4: register [0x4000,0x4007] min 2, exclusion, zext|inttype —
    // cc:389 zext wins over cc:391 inttype
    stds.entry.push_back(makeEntry(4, reg, 0x4000, 8, 2, 0,
                                   ParamEntry::smallsize_zext |
                                   ParamEntry::smallsize_inttype));
    // e5: register [0x5000,0x5007] min 2, exclusion, inttype|sext —
    // cc:391 inttype wins over cc:393 sext
    stds.entry.push_back(makeEntry(5, reg, 0x5000, 8, 2, 0,
                                   ParamEntry::smallsize_inttype |
                                   ParamEntry::smallsize_sext));
    stds.numgroup = 6;
    struct AeCase { const char *spcname; AddrSpace *spc; uintb off; int4 sz; };
    const AeCase ae[] = {
      { "reg", reg, 0x100, 2 },    // justified -> ZEXT, container 0x100/8
      { "reg", reg, 0x108, 2 },    // second slot -> ZEXT, container 0x108/8
      { "reg", reg, 0x102, 2 },    // unjustified (2 % 8) -> COPY
      { "stack", stack, 0x100, 2 }, // foreign space at numeric hit -> COPY
      { "reg", reg, 0x100, 8 },    // sz >= alignment (and all later gates)
      { "reg", reg, 0x100, 1 },    // list minSize skip (e0/e1/e2 min 2)
      { "reg", reg, 0x200, 2 },    // join guard cc:376 (low piece justifies)
      { "ram", ram, 0x2000, 4 },   // exclusion container 0x2000/16 -> PIECE
      { "ram", ram, 0x2002, 4 },   // unjustified (offset 2) -> COPY
      { "stack", stack, 0x2000, 4 }, // address.cc:133 route -> COPY
      { "reg", reg, 0x3000, 2 },   // sext only -> INT_SEXT, container 0x3000/8
      { "reg", reg, 0x4000, 2 },   // zext|inttype -> INT_ZEXT (cc:389 first)
      { "reg", reg, 0x5000, 2 },   // inttype|sext -> PIECE (cc:391 before 393)
    };
    for (const AeCase &c : ae)
      printAe(&stds, c.spcname, c.spc, c.off, c.sz);
  }

  return 0;
}

File: tests/oracle/fspec_possibleparam_1204.rs
// FSPEC-POSSIBLEPARAM-1204: Rugra comparand for the locked Ghidra
// 12.0.4 oracle (FSPEC-POSSIBLEPARAM-JOIN-0006, A46 residual 3).
// Mirrors tests/oracle/fspec_possibleparam_1204.cc case for case:
//  - plain_out_entries: ParamListStandardOut::possible_param iterates
//    the raw entry list (no resolver window, no caller-level space
//    filter, no minSize gate) and accepts any justifiedContain >= 0
//    (fspec.cc:1765-1774); space rejection happens only inside
//    justifiedContain (fspec.cc:269 / address.cc:133).
//  - join_out_entries: join entries are reachable through the
//    per-piece join walk (fspec.cc:251-262): 0x104/4 gives offset 4
//    >= 0 -> true, unlike find_entry(just=true)'s == 0 gate.
//  - cross_space_join: the foreign low piece contributes
//    address.cc:133 -1 (+4 skip), the containing high piece cur=0,
//    so a ram query returns offset 4 >= 0 -> true.
//  - assumed_extension: ParamListStandard::assumed_extension
//    (fspec.cc:1426) -> ParamEntry::assumed_extension (fspec.cc:366)
//    with the space-aware cc:377 gate, the cc:376 join guard, the sz
//    gates, the list minSize skip, and both container shapes.

use rugra::address::Address;
use rugra::fspec::{param_entry_flags, ParamEntry, ParamListStandard,
                   ParamListStandardOut, TypeClass, VarnodeData};
use rugra::space::AddressSpace;

// staged-loader entry builder (the pub equivalent of the C++ fixture's
// direct field writes; same pattern as the fspec_findentry_1204
// fixture).
fn make_entry(
    grp: i32,
    spc: AddressSpace,
    base: u64,
    size: i32,
    min_size: i32,
    alignment: i32,
    flags: u32,
) -> ParamEntry {
    let mut e = ParamEntry::new(grp);
    *e.flags_mut() = flags;
    e.set_type_class(TypeClass::General);
    e.set_space(spc);
    e.set_base(base);
    e.set_sizes(size, min_size);
    e.set_alignment(alignment); // normalizes alignment == size to 0
    e
}

// Stage a join ParamEntry (space = join space + pieces, most
// significant first) — the state resolveJoin caches after decoding a
// join <pentry>.
fn make_join_entry(
    grp: i32,
    size: i32,
    min_size: i32,
    pieces: &[VarnodeData],
) -> ParamEntry {
    let mut e = ParamEntry::new(grp);
    e.set_type_class(TypeClass::General);
    e.set_space(AddressSpace::Join);
    e.set_base(0);
    e.set_sizes(size, min_size);
    e.set_alignment(0);
    e.set_join_pieces(pieces.to_vec());
    e
}

fn print_pp(out: &ParamListStandardOut, spcname: &str, space: AddressSpace,
            off: u64, sz: i32) {
    let hit = out.possible_param(space, Address::new(off), sz);
    println!("  pp spc={} off=0x{:x} sz={} -> {}", spcname, off, sz,
             if hit { "true" } else { "false" });
}

fn print_ae(model: &ParamListStandard, spcname: &str, space: AddressSpace,
            off: u64, sz: i32) {
    let mut res = VarnodeData { space: AddressSpace::Ram, offset: 0, size: 0 };
    let ext = model.assumed_extension(space, Address::new(off), sz, &mut res);
    let mut line = format!("  ae spc={} off=0x{:x} sz={} -> {}", spcname, off,
                           sz, ext.name());
    if ext != rugra::opcodes::OpCode::CPUI_COPY {
        line.push_str(&format!(" res={}:0x{:x}/{}", res.space.name(),
                               res.offset, res.size));
    }
    println!("{}", line);
}

fn main() {
    println!("schema=1|fixture=FSPEC-POSSIBLEPARAM-1204|oracle=e40ed13014025f82488b1f8f7bca566894ac376b");

    // ---- case 1: plain output entries --------------------------------
    println!("case=plain_out_entries");
    {
        let mut out = ParamListStandardOut::new();
        // e0: register [0x200,0x207] min 4, exclusion, force-left
        out.base.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x200, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        // e1: register [0x1000,0x100F] min 4, alignment 8 (aligned route)
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x1000, 16, 4, 8, 0,
        ));
        // e2: ram [0x2000,0x200F] min 2, exclusion, force-left
        out.base.entry_mut().push(make_entry(
            2, AddressSpace::Ram, 0x2000, 16, 2, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        out.base.set_num_group(3);
        let pp: [(&str, AddressSpace, u64, i32); 10] = [
            ("reg", AddressSpace::Register, 0x200, 4),    // justified hit -> true
            ("reg", AddressSpace::Register, 0x202, 2),    // unjustified offset 2: >= 0 -> true
            ("reg", AddressSpace::Register, 0x208, 4),    // above e0, below e1 -> false
            ("reg", AddressSpace::Register, 0x1000, 8),   // aligned justified hit -> true
            ("reg", AddressSpace::Register, 0x1002, 4),   // aligned unjustified (2) -> true
            ("stack", AddressSpace::Stack, 0x1000, 8),    // cc:269 foreign-space guard -> false
            ("stack", AddressSpace::Stack, 0x200, 4),     // address.cc:133 route -> false
            ("ram", AddressSpace::Ram, 0x2000, 1),        // NO minSize gate (min 2 > 1) -> true
            ("reg", AddressSpace::Register, 0x200, 1),    // NO minSize gate (min 4 > 1) -> true
            ("unique", AddressSpace::Unique, 0x200, 4),   // foreign space -> false
        ];
        for &(name, space, off, sz) in pp.iter() {
            print_pp(&out, name, space, off, sz);
        }
    }

    // ---- case 2: join entries ARE reachable --------------------------
    println!("case=join_out_entries");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Register, offset: 0x104, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x100, size: 4 },
        ];
        let mut out = ParamListStandardOut::new();
        // e0: join reg:0x104 (high) + reg:0x100 (low)
        out.base.entry_mut().push(make_join_entry(0, 8, 4, &pieces));
        // e1: plain register [0x200,0x207] min 4
        out.base.entry_mut().push(make_entry(
            1, AddressSpace::Register, 0x200, 8, 4, 0,
            param_entry_flags::FORCE_LEFT_JUSTIFY,
        ));
        out.base.set_num_group(2);
        let pp: [(&str, AddressSpace, u64, i32); 6] = [
            ("reg", AddressSpace::Register, 0x100, 4),    // low piece offset 0 -> true (join!)
            ("reg", AddressSpace::Register, 0x104, 4),    // walk offset 4 >= 0 -> true (join!)
            ("reg", AddressSpace::Register, 0x100, 8),    // join walk -1, e1 no -> false
            ("reg", AddressSpace::Register, 0x102, 4),    // pokes out of both pieces -> false
            ("reg", AddressSpace::Register, 0x200, 4),    // plain entry hit -> true
            ("stack", AddressSpace::Stack, 0x100, 4),     // per-piece address.cc:133 -> false
        ];
        for &(name, space, off, sz) in pp.iter() {
            print_pp(&out, name, space, off, sz);
        }
    }

    // ---- case 3: cross-space join pieces ------------------------------
    println!("case=cross_space_join");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Ram, offset: 0x200, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x200, size: 4 },
        ];
        let mut out = ParamListStandardOut::new();
        out.base.entry_mut().push(make_join_entry(0, 8, 4, &pieces));
        out.base.set_num_group(1);
        let pp: [(&str, AddressSpace, u64, i32); 5] = [
            ("reg", AddressSpace::Register, 0x200, 4),    // low piece offset 0 -> true
            ("ram", AddressSpace::Ram, 0x200, 4),         // foreign low -1 (+4), high cur=0 -> 4 >= 0 -> true
            ("reg", AddressSpace::Register, 0x204, 4),    // outside both pieces -> false
            ("reg", AddressSpace::Register, 0x200, 8),    // low pokes out, high foreign -> -1 -> false
            ("stack", AddressSpace::Stack, 0x200, 4),     // both pieces foreign -> false
        ];
        for &(name, space, off, sz) in pp.iter() {
            print_pp(&out, name, space, off, sz);
        }
    }

    // ---- case 4: assumed_extension gates and containers ----------------
    println!("case=assumed_extension");
    {
        let pieces = [
            VarnodeData { space: AddressSpace::Register, offset: 0x204, size: 4 },
            VarnodeData { space: AddressSpace::Register, offset: 0x200, size: 4 },
        ];
        let mut model = ParamListStandard::new();
        // e0: register [0x100,0x11F] min 2 alignment 8, smallsize zext
        model.entry_mut().push(make_entry(
            0, AddressSpace::Register, 0x100, 32, 2, 8,
            param_entry_flags::SMALLSIZE_ZEXT,
        ));
        // e1: join reg:0x204 + reg:0x200 min 2, smallsize sext (never
        // extends: cc:376 join guard)
        model.entry_mut().push(make_join_entry(1, 8, 2, &pieces));
        *model.entry_mut().last_mut().unwrap().flags_mut() =
            param_entry_flags::SMALLSIZE_SEXT;
        // e2: ram [0x2000,0x200F] min 2, exclusion, smallsize inttype
        model.entry_mut().push(make_entry(
            2, AddressSpace::Ram, 0x2000, 16, 2, 0,
            param_entry_flags::SMALLSIZE_INTTYPE,
        ));
        // e3: register [0x3000,0x3007] min 2, exclusion, smallsize sext
        model.entry_mut().push(make_entry(
            3, AddressSpace::Register, 0x3000, 8, 2, 0,
            param_entry_flags::SMALLSIZE_SEXT,
        ));
        // e4: register [0x4000,0x4007] min 2, exclusion, zext|inttype —
        // cc:389 zext wins over cc:391 inttype
        model.entry_mut().push(make_entry(
            4, AddressSpace::Register, 0x4000, 8, 2, 0,
            param_entry_flags::SMALLSIZE_ZEXT
                | param_entry_flags::SMALLSIZE_INTTYPE,
        ));
        // e5: register [0x5000,0x5007] min 2, exclusion, inttype|sext —
        // cc:391 inttype wins over cc:393 sext
        model.entry_mut().push(make_entry(
            5, AddressSpace::Register, 0x5000, 8, 2, 0,
            param_entry_flags::SMALLSIZE_INTTYPE
                | param_entry_flags::SMALLSIZE_SEXT,
        ));
        model.set_num_group(6);
        let ae: [(&str, AddressSpace, u64, i32); 13] = [
            ("reg", AddressSpace::Register, 0x100, 2),    // justified -> ZEXT, container 0x100/8
            ("reg", AddressSpace::Register, 0x108, 2),    // second slot -> ZEXT, container 0x108/8
            ("reg", AddressSpace::Register, 0x102, 2),    // unjustified (2 % 8) -> COPY
            ("stack", AddressSpace::Stack, 0x100, 2),     // foreign space at numeric hit -> COPY
            ("reg", AddressSpace::Register, 0x100, 8),    // sz >= alignment (and all later gates)
            ("reg", AddressSpace::Register, 0x100, 1),    // list minSize skip (e0/e1/e2 min 2)
            ("reg", AddressSpace::Register, 0x200, 2),    // join guard cc:376 (low piece justifies)
            ("ram", AddressSpace::Ram, 0x2000, 4),        // exclusion container 0x2000/16 -> PIECE
            ("ram", AddressSpace::Ram, 0x2002, 4),        // unjustified (offset 2) -> COPY
            ("stack", AddressSpace::Stack, 0x2000, 4),    // address.cc:133 route -> COPY
            ("reg", AddressSpace::Register, 0x3000, 2),   // sext only -> INT_SEXT, container 0x3000/8
            ("reg", AddressSpace::Register, 0x4000, 2),   // zext|inttype -> INT_ZEXT (cc:389 first)
            ("reg", AddressSpace::Register, 0x5000, 2),   // inttype|sext -> PIECE (cc:391 before 393)
        ];
        for &(name, space, off, sz) in ae.iter() {
            print_ae(&model, name, space, off, sz);
        }
    }
}
